File: 2048-flutter/lib/models/game.dart
import 'dart:math';
import 'tile.dart';

enum Direction { up, down, left, right }

class Game {
  static const int size = 4;
  static const int winningValue = 2048;
  final List<List<Tile?>> grid;
  int score = 0;
  bool gameOver = false;
  bool won = false;

  Game() : grid = List.generate(
    size,
    (i) => List.generate(size, (j) => null),
  ) {
    addRandomTile();
    addRandomTile();
  }

  void addRandomTile() {
    List<List<int>> emptyCells = [];
    for (int i = 0; i < size; i++) {
      for (int j = 0; j < size; j++) {
        if (grid[i][j] == null) {
          emptyCells.add([i, j]);
        }
      }
    }

    if (emptyCells.isEmpty) return;

    final random = Random();
    final cell = emptyCells[random.nextInt(emptyCells.length)];
    final value = random.nextDouble() < 0.9 ? 2 : 4;
    grid[cell[0]][cell[1]] = Tile(
      value: value,
      row: cell[0],
      col: cell[1],
      isNew: true,
    );
  }

  bool move(Direction direction) {
    if (gameOver || won) return false;

    bool moved = false;
    List<List<Tile?>> newGrid = List.generate(
      size,
      (i) => List.generate(size, (j) => null),
    );

    switch (direction) {
      case Direction.left:
        for (int row = 0; row < size; row++) {
          List<Tile?> currentRow = grid[row].where((tile) => tile != null).toList();
          List<Tile?> mergedRow = _mergeTiles(currentRow);
          for (int col = 0; col < mergedRow.length; col++) {
            if (mergedRow[col] != null) {
              newGrid[row][col] = mergedRow[col]!.copyWith(row: row, col: col);
            }
          }
        }
        break;
      case Direction.right:
        for (int row = 0; row < size; row++) {
          List<Tile?> currentRow = grid[row].where((tile) => tile != null).toList().reversed.toList();
          List<Tile?> mergedRow = _mergeTiles(currentRow);
          for (int col = 0; col < mergedRow.length; col++) {
            if (mergedRow[col] != null) {
              newGrid[row][size - 1 - col] = mergedRow[col]!.copyWith(row: row, col: size - 1 - col);
            }
          }
        }
        break;
      case Direction.up:
        for (int col = 0; col < size; col++) {
          List<Tile?> currentCol = [];
          for (int row = 0; row < size; row++) {
            if (grid[row][col] != null) currentCol.add(grid[row][col]);
          }
          List<Tile?> mergedCol = _mergeTiles(currentCol);
          for (int row = 0; row < mergedCol.length; row++) {
            if (mergedCol[row] != null) {
              newGrid[row][col] = mergedCol[row]!.copyWith(row: row, col: col);
            }
          }
        }
        break;
      case Direction.down:
        for (int col = 0; col < size; col++) {
          List<Tile?> currentCol = [];
          for (int row = size - 1; row >= 0; row--) {
            if (grid[row][col] != null) currentCol.add(grid[row][col]);
          }
          List<Tile?> mergedCol = _mergeTiles(currentCol);
          for (int row = 0; row < mergedCol.length; row++) {
            if (mergedCol[row] != null) {
              newGrid[size - 1 - row][col] = mergedCol[row]!.copyWith(
                row: size - 1 - row,
                col: col,
              );
            }
          }
        }
        break;
    }

    moved = !_gridsAreEqual(grid, newGrid);
    if (moved) {
      grid.clear();
      grid.addAll(newGrid);
      addRandomTile();
      _checkGameStatus();
    }

    return moved;
  }

  List<Tile?> _mergeTiles(List<Tile?> tiles) {
    List<Tile?> merged = List.filled(size, null);
    int index = 0;
    int? lastValue;
    int lastIndex = -1;

    for (Tile? tile in tiles) {
      if (tile == null) continue;

      if (lastValue == tile.value) {
        merged[lastIndex] = Tile(
          value: lastValue! * 2,
          row: tile.row,
          col: tile.col,
          isMerged: true,
        );
        score += lastValue * 2;
        if (lastValue * 2 == winningValue) {
          won = true;
        }
        lastValue = null;
      } else {
        lastValue = tile.value;
        lastIndex = index;
        merged[index] = tile;
        index++;
      }
    }

    return merged;
  }

  bool _gridsAreEqual(List<List<Tile?>> grid1, List<List<Tile?>> grid2) {
    for (int i = 0; i < size; i++) {
      for (int j = 0; j < size; j++) {
        if ((grid1[i][j] == null) != (grid2[i][j] == null)) return false;
        if (grid1[i][j] != null && grid2[i][j] != null) {
          if (grid1[i][j]!.value != grid2[i][j]!.value) return false;
        }
      }
    }
    return true;
  }

  void _checkGameStatus() {
    if (won) return;

    bool canMove = false;
    for (int i = 0; i < size; i++) {
      for (int j = 0; j < size; j++) {
        if (grid[i][j] == null) {
          canMove = true;
          break;
        }
        if (i < size - 1 && grid[i][j]?.value == grid[i + 1][j]?.value) {
          canMove = true;
          break;
        }
        if (j < size - 1 && grid[i][j]?.value == grid[i][j + 1]?.value) {
          canMove = true;
          break;
        }
      }
      if (canMove) break;
    }
    gameOver = !canMove;
  }

  void reset() {
    grid.clear();
    grid.addAll(List.generate(
      size,
      (i) => List.generate(size, (j) => null),
    ));
    score = 0;
    gameOver = false;
    won = false;
    addRandomTile();
    addRandomTile();
  }
}

File: 2048-vanilla/js/game.js
// 游戏核心逻辑
export class Game {
    constructor(size = 4) {
        this.size = size;
        this.init();
    }

    init() {
        this.grid = Array(this.size).fill().map(() => Array(this.size).fill(0));
        this.score = 0;
        this.gameOver = false;
        this.mergedTiles = [];
        this.newTiles = [];
        this.addRandomTile();
        this.addRandomTile();
        this.updateView();
    }

    addRandomTile() {
        const emptyCells = [];
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.grid[i][j] === 0) {
                    emptyCells.push({x: i, y: j});
                }
            }
        }
        if (emptyCells.length) {
            const {x, y} = emptyCells[Math.floor(Math.random() * emptyCells.length)];
            this.grid[x][y] = Math.random() < 0.9 ? 2 : 4;
            this.newTiles.push({x, y});
        }
    }

    move(direction) {
        if (this.gameOver) return false;

        this.newTiles = [];
        this.mergedTiles = [];
        const prevGrid = JSON.stringify(this.grid);
        
        switch(direction) {
            case 'up': this.moveUp(); break;
            case 'down': this.moveDown(); break;
            case 'left': this.moveLeft(); break;
            case 'right': this.moveRight(); break;
        }

        const moved = prevGrid !== JSON.stringify(this.grid);
        if (moved) {
            this.addRandomTile();
            this.updateView();
            this.checkGameOver();
        }
        return moved;
    }

    moveLeft() {
        let moved = false;
        for (let i = 0; i < this.size; i++) {
            const row = this.grid[i];
            const newRow = new Array(this.size).fill(0);
            let writePos = 0;
            let readPos = 0;
            let lastMerged = false;

            // 移动和合并
            while (readPos < this.size) {
                if (row[readPos] === 0) {
                    readPos++;
                    continue;
                }

                if (writePos === 0 || newRow[writePos - 1] === 0) {
                    newRow[writePos] = row[readPos];
                    if (writePos !== readPos) {
                        moved = true;
                    }
                    lastMerged = false;
                    writePos++;
                } else if (!lastMerged && newRow[writePos - 1] === row[readPos]) {
                    newRow[writePos - 1] *= 2;
                    this.score += newRow[writePos - 1];
                    this.mergedTiles.push({x: i, y: writePos - 1});
                    moved = true;
                    lastMerged = true;
                } else {
                    newRow[writePos] = row[readPos];
                    if (writePos !== readPos) {
                        moved = true;
                    }
                    lastMerged = false;
                    writePos++;
                }
                readPos++;
            }

            if (moved) {
                this.grid[i] = newRow;
            }
        }
        return moved;
    }

    moveRight() {
        // 先反转网格，使用 moveLeft，再反转回来
        this.grid = this.grid.map(row => row.reverse());
        const moved = this.moveLeft();
        this.grid = this.grid.map(row => row.reverse());
        // 更新合并方块的位置
        this.mergedTiles = this.mergedTiles.map(({x, y}) => ({
            x: x,
            y: this.size - 1 - y
        }));
        return moved;
    }

    moveUp() {
        this.rotateGrid();
        const moved = this.moveLeft();
        this.rotateGrid(3);
        return moved;
    }

    moveDown() {
        this.rotateGrid();
        const moved = this.moveRight();
        this.rotateGrid(3);
        return moved;
    }

    rotateGrid(times = 1) {
        for (let t = 0; t < times; t++) {
            const newGrid = Array(this.size).fill().map(() => Array(this.size).fill(0));
            for (let i = 0; i < this.size; i++) {
                for (let j = 0; j < this.size; j++) {
                    newGrid[j][this.size - 1 - i] = this.grid[i][j];
                }
            }
            this.grid = newGrid;
            
            // 更新合并和新增方块的位置
            this.mergedTiles = this.mergedTiles.map(({x, y}) => ({
                x: y,
                y: this.size - 1 - x
            }));
            this.newTiles = this.newTiles.map(({x, y}) => ({
                x: y,
                y: this.size - 1 - x
            }));
        }
    }

    updateView() {
        const gameGrid = document.getElementById('game-grid');
        gameGrid.innerHTML = '';
        gameGrid.style.gridTemplateColumns = `repeat(${this.size}, 1fr)`;

        // 计算单个格子的大小
        const containerWidth = gameGrid.offsetWidth;
        const spacing = parseInt(getComputedStyle(document.documentElement).getPropertyValue('--grid-spacing'));
        const cellSize = (containerWidth - (this.size + 1) * spacing) / this.size;
        
        // 更新CSS变量
        document.documentElement.style.setProperty('--cell-size', cellSize + 'px');

        // 创建背景网格
        for (let i = 0; i < this.size * this.size; i++) {
            const cell = document.createElement('div');
            cell.className = 'grid-cell';
            gameGrid.appendChild(cell);
        }

        // 添加数字块
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.grid[i][j] !== 0) {
                    const tile = document.createElement('div');
                    tile.className = `tile tile-${this.grid[i][j]}`;
                    tile.textContent = this.grid[i][j];
                    
                    // 计算位置
                    const x = j * (cellSize + spacing) + spacing;
                    const y = i * (cellSize + spacing) + spacing;
                    
                    tile.style.transform = `translate(${x}px, ${y}px)`;
                    
                    // 添加动画类
                    const isNew = this.newTiles.some(t => t.x === i && t.y === j);
                    const isMerged = this.mergedTiles.some(t => t.x === i && t.y === j);
                    
                    if (isNew) {
                        tile.classList.add('new');
                    } else if (isMerged) {
                        tile.classList.add('merged');
                    }
                    
                    gameGrid.appendChild(tile);
                }
            }
        }

        // 更新分数
        const scoreElement = document.getElementById('score');
        if (scoreElement) {
            scoreElement.textContent = this.score;
        }
    }

    checkGameOver() {
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.grid[i][j] === 0) return false;
                if (i < this.size - 1 && this.grid[i][j] === this.grid[i + 1][j]) return false;
                if (j < this.size - 1 && this.grid[i][j] === this.grid[i][j + 1]) return false;
            }
        }
        this.gameOver = true;
        document.getElementById('game-over').style.display = 'flex';
        return true;
    }

    getGrid() {
        return this.grid;
    }

    getScore() {
        return this.score;
    }

    isGameOver() {
        return this.gameOver;
    }

    setSize(size) {
        this.size = size;
        this.init();
    }
}
